### ISSUE TITLE
twisted.web.static.File should compose a FilePath-like object rather than inheriting from FilePath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3765]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3765|
|Type|enhancement|
|Created|2009-04-08 04:28:05Z|
|Branch|https://github.com/twisted/twisted/tree/static-path-3765|

Right now `twisted.web.static.File` subclasses from FilePath.  Among other things, this means:

  * it's unnecessarily difficult to test, because you have to talk to a real filesystem
  * you can't serve files out of a zip file
  * you can't serve files out of an in-memory filesystem (although `MemoryPath` is a separate issue)
  * subclassing `File` to customize it is more difficult than it ought to be because all the responsibilities are mixed together.

I suggest we create a new class, `twisted.web.static.Path`, which does what `File` does, but instead of mixing in `FilePath`, takes a `FilePath` (or really an `IFilePath` provider) to its constructor.

<details><summary>Searchable metadata</summary>

```
trac-id__3765 3765
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_static_path_3765 branches/static-path-3765
branch_author__rwall rwall
status__new new
resolution__None None
component__web web
keywords__tape tape
time__1239164885000000 1239164885000000
changetime__1398023062000000 1398023062000000
version__None None
owner__rwall rwall
cc__jack cc__exarkun cc__Julian
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

What is the _tape_ keyword?

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|jack commented|
|-|-|

Glyph told me to add it to all the bugs I filed related to my implemented of Tape on top of twisted.web.  Tape is at [here](http://github.com/metajack/tape/tree/master) if you are curious.  The email thread where the discussion happened is on twisted-web.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj commented|
|-|-|

(In [39809]) Branching to 'static-path-3765'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj commented|
|-|-|

Before I go any further, I'd like a design review of the approach in log:branches/static-path-3765

 * I've created a new Path resource -- in which I have deliberately only added traversal handling
 * I've then added what I hope will be flexible renderers for files and directories -- these can be reused by people wanting to serve a static file or directory from their own Renderers Eg a video serving renderer which first looks up the requested video from the query string and instantiates a FilePathRenderer which takes care of encoding and hooking up a producer.
 * I've updated DirectoryLister to accept a FilePath -- this could be broken off into another ticket -- or I could write a new Directory lister from scratch which uses twistd.web template.
 * I've added two new demo functions and tweaked t.w.tap so that they can be called using a combination of --class and --path
 * Also raised [#6727](https://github.com/twisted/twisted/issues/6727) which makes the traversal function uglier than it needs to be.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set owner to @wallrj|
|-|-|

Replying to [rwall](#note_7):
> Before I go any further, I'd like a design review of the approach in log:branches/static-path-3765
> 

OK, I've had a look.

>  * I've created a new Path resource -- in which I have deliberately only added traversal handling

I really like the way you've parameterized not just the `FilePath`-ness, but all of the class's behaviors.

Naming-wise, I believe this ought to be in a new module; let's say, `twisted.web.path.Path`.  This is definitively _not_ for "static content".  `FilePathRenderer(path)` might be, although I'd actually name that (as currently conceived) differently: `twisted.web.static.ContentsOfPath(path)` (since that truly _is_ about static content now!).

The one thing that bothers me a little bit about its traversal implementation is the lack of the somewhat-obscure "registry" object that `FilePath` objects previously shared.  It's necessary for the cache() directive in RPY files to work properly.  However, if you're going to be doing stuff over with a *good* design, you will probably want to remove the fact that it's inheriting from `Componentized`; it's really for caching paths, and the `Componentized` bit is just there to help you randomly share global state between your `.rpy` resources.

Experience has taught us in the intervening years since `static.File` was desgined that the right way to share state between resources in a deployed application is to just have some instance variables.  For more significant sharing, like, for example, a web application that wants to access the strcred-configured checkers, we'd need something non-web-related anyway.  And there are enough ways to share random global state in Python, we probably don't need to add more.

>  * I've then added what I hope will be flexible renderers for files and directories -- these can be reused by people wanting to serve a static file or directory from their own Renderers Eg a video serving renderer which first looks up the requested video from the query string and instantiates a FilePathRenderer which takes care of encoding and hooking up a producer.

Although the _factoring_ of the current `static.File` is atrocious, the _functionality_ is somewhat desirable.
It's nice to be able to dispatch on file extension and render various types of static or dynamic content; this vastly simplifies application deployment.
This is why I suggested re-naming the `FilePathRenderer` class; its job ought to be simple: just spit out static content.  However, the default `_fileRenderer` ought to do a bit more than that.
The logic which is currently embedded in `processors.get(extension)` ought to be an _instance_ of a separate, dedicated class, `ExtensionMapResourceFactory`, let's say, which maps extensions to resource factories.  (Or perhaps a bound method on an instance of that class.)
Similarly, the documentation should emphasize the contract with the object ("1-argument callable returning an `IResource` provider") and not say anything about classes.  You don't need a class anywhere to have a useful resouce factory :).
And speaking of documentation: some narrative documentation might help to elucidate the design, even if you still don't want a full code review after this.

(Also, by the way, the "is your filesystem case-insensitive" check is something that ought to be handled by the filesystem layer, i.e. `FilePath` itself.  Just asking if the user is on Windows is far from sufficient: you can mount case-insensitive filesystems on every platform.  My inclination would be to just always be case-insensitive by default in the new implementation.)

>  * I've updated DirectoryLister to accept a FilePath -- this could be broken off into another ticket -- or I could write a new Directory lister from scratch which uses twistd.web template.

Oh, please do the latter (maybe in a separate ticket).  That was kinda the point of t.w.t, really.

If you write a new directory lister that accepts a `FilePath`, then you can just pass it either a `File` (since it inherits `FilePath`) or a `Path`'s `IFilePath`.

>  * Also raised [#6727](https://github.com/twisted/twisted/issues/6727) which makes the traversal function uglier than it needs to be.

I don't see how this function would be nicer with different `ZipPath` behavior.  `""` is special to this dynamic path behavior because it indicates an index: that will be true in any case.

Thanks for working on this, I look forward to it landing.  Was this the sort of review you were looking for?

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_9"></a>|@wallrj commented|
|-|-|

See ticket:6951#comment:5 which has some changes to the static.File ForbiddenResource behaviour.

And thanks Glyph for that review, it was exactly the sort of feedback that I wanted. Just not sure when I'll get round to finishing this off.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Happy to hear that there is work for a new static file resource which uses isolated/component file access.

Just some feedback from using the current static.File code.

I am no longer using twisted.web.static.File for url traversal and folder listing.
I still use it as a 'leaf' resource.

Beside custom opening of the file, I was also looking for impersonated file access.

This was ok since isdir(), exists(), openForReading() methods were already there... but I had to scan to code and make sure no other methods are called... so I would have to overwrite the code in many places.

So, it is nice that the new static.Path can get the filesystem access as a component so that I can pass/inject my InMemoryPath or ImpersoantedPath or ZipPath implementation. 

In _pathNotFoundRenderer I also wanted to get the name of the child which is not found and parent filePath. This could allow extending the _pathNotFoundRenderer to implement file upload using a POST/PUT request.

I am not sure if is ok to call them Renderers since they are returned by getChildWithDefault and they should be proper resources.

Why are the new Path and FilePathRenderer not inheriting from Resource and implementing getChildWithDefault and not getChild ?

Also, I prefer to have dedicated calls to render_GET... render_HEAD, in this way it should be easier to add render_POST or render_PUT handlers.

Beside read only access, I also need file management support to create files, folders, delete files, rename files.
Not sure if this is wanted ... or users looking for this kind of things should use completely different objects.

I found twisted web templates to be too simple and also slower than Jinja2.. so It would be nice to not make it harder for users to implement their own rendering.
I have implementing what I called a ViewModel, which is similar web.template.Element as it is contains only data and is the only thing passed to the template to fill the template.
Not sure if this comment is of any help.

Here is my URL traversal code https://gist.github.com/adiroiban/8998281

I have also reduced filesystem calls by accumulating all path segments and doing the filesystem access only once after all URI parts were traversed.
Right now the Static url traversal stops at each path/uri part/segment, calls stats.
Maybe this is a stupid thing to do.

I assume that this ticket will get huge... so maybe it requires a parent ticket focused on a simple new resource designed to do URL traversal and dispatching of File or Directory resources.
It will delegated the final file rendering to the existing static.File and directory
rendering to existing directory listing.
It should extract the url traversal part out of current static.File and should be designed to be assigned as a root resource.

Then this ticket can focus on file opening and file reading part.

Another ticket could focus in directory listing rendering part.

I am still new to twisted.web ... so maybe I am doing something wrong and there are already simple ways of doing what I want.
Creating my own resources was very easy but I am thinking that maybe we can have some common code in twisted.web as a starting point or reference for more complicated/customized usage.

Thanks!
